fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentHandlerFactory.java

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.http.api;
+
+import static org.fcrepo.kernel.api.RdfLexicon.EXTERNAL_CONTENT;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.fcrepo.kernel.api.exception.ExternalMessageBodyException;
+import org.slf4j.Logger;
+
+/**
+ * Constructs ExternalContentHandler objects from link headers
+ *
+ * @author bbpennel
+ */
+public class ExternalContentHandlerFactory {
+
+    private static final Logger LOGGER = getLogger(ExternalContentHandlerFactory.class);
+
+    private ExternalContentPathValidator validator;
+
+    /**
+     * Looks for ExternalContent link header and if it finds one it will return a new ExternalContentHandler object
+     * based on the found Link header. If multiple external content headers were found or the URI provided in the
+     * header is not a valid external content path, then an ExternalMessageBodyException will be thrown.
+     *
+     * @param links links from the request header
+     * @return External Content Handler Object if Link header found, else null
+     * @throws ExternalMessageBodyException thrown if more than one external content link was provided, or if the URL
+     *         of the header was not a valid external content path.
+     */
+    public ExternalContentHandler createFromLinks(final List<String> links) throws ExternalMessageBodyException {
+        if (links == null) {
+            return null;
+        }
+
+        final List<String> externalContentLinks = links.stream()
+                .filter(x -> x.contains(EXTERNAL_CONTENT.toString()))
+                .collect(Collectors.toList());
+
+        if (externalContentLinks.size() > 1) {
+            // got a problem, you can only have one ExternalContent links
+            throw new ExternalMessageBodyException("More then one External Content Link header in request");
+        } else if (externalContentLinks.size() == 1) {
+            final String link = externalContentLinks.get(0);
+            final String uri = getUriString(link);
+            // Validate that the URI is valid according to allowed set of external uris
+            try {
+                validator.validate(uri);
+            } catch (final ExternalMessageBodyException e) {
+                LOGGER.warn("Rejected invalid external path {}", uri);
+                throw e;
+            }
+
+            return new ExternalContentHandler(link);
+        }
+
+        return null;
+    }
+
+    private static String getUriString(final String link) {
+        final String value = link.trim();
```

Comment:
is there a reason to not use Link here? sort of like-so: https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentHandler.java#L112

Reply:
Yes, I had to do this processing before it went through the `Link` class because that class automatically evaluates path modifiers. So if I get a url of `file:///path/to/stuff/../file.txt` and put it through the `Link` class and call .getURL on it, the value comes back as `file:///path/to/file.txt`. This is the reason I had to change when validation happened from operating on the ExternalContentHandler to happening before the handler.

Reply:
Ah, okay.  Thanks for the explanation. :)